module/spl/spl-vnode.c

```diff
@@ -235,9 +235,17 @@ vn_rdwr(uio_rw_t uio, vnode_t *vp, void *addr, ssize_t len, offset_t off,
         set_fs(get_ds());
 
 	if (uio & UIO_WRITE)
+#ifdef HAVE_VFS_WRITE
 		rc = vfs_write(fp, addr, len, &offset);
+#else
+		rc = kernel_write(fp, addr, len, &offset);
+#endif /* HAVE_VFS_WRITE */
 	else
+#ifdef HAVE_VFS_WRITE
 		rc = vfs_read(fp, addr, len, &offset);
+#else
+		rc = kernel_read(fp, addr, len, &offset);
+#endif /* HAVE_VFS_WRITE */
```

Comment:
It looks like `kernel_write()` was exported in v3.8 torvalds/linux@7bb307e894d51308aa0582a8c4cc5875bbc645b9 and `kernel_read()` in 3.9 torvalds/linux@3dc20cb282ec03cc4c997130d680c800011ed479.  This is arguably when we should have moved over to these new interfaces.

What we need to do handle this is to add a configure check for both `kernel_read` and `kernel_write` and a compatibility wrapper for each of these called `spl_kernel_read()` and `spl_kernel_write()` for consistency.  The wrappers should prefer the `kernel_*` interfaces if they're determined to be available.  And since these functions are only called from `spl-vnode.c` they can be added there and declared static.

There are a couple things to be careful of when implementing the wrappers.

* The arguments for the functions do not appear in the same order.
* `vfs_read/write` functions don't handle the `set_fs()` bit but the `kernel_read/write` functions do.  This logic will need to be pulled in to the wrapper for only the `vfs_read/write` case and out of `vn_rdwr`.

Reply:
Unfortunately, I won't have the time to work further on this issue. Concerning your comments: I would actually leave it the way it is. All previous ZFS versions compiled against kernels < 4.14 have been using vfs_read/write and were tested against this. Why risk catching a regression with such kernels due to a possibly buggy early implemenation of kernel_read/write? I don't see a benefit, only potential problems when using kernel_read/write with kernels < 4.14.
I signed off the commit rebased and pushed.